File: buildSrc/src/main/groovy/org/stepik/gradle/plugins/jetbrains/BasePlugin.java
package org.stepik.gradle.plugins.jetbrains;

import org.gradle.api.Action;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.file.DuplicatesStrategy;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.internal.jvm.Jvm;
import org.gradle.language.jvm.tasks.ProcessResources;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.stepik.gradle.plugins.jetbrains.dependency.DependencyManager;
import org.stepik.gradle.plugins.jetbrains.dependency.ProductDependency;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * @author meanmail
 */
public abstract class BasePlugin implements Plugin<Project> {
    private static final Logger LOG = LoggerFactory.getLogger(BasePlugin.class);

    protected String extensionName;
    protected String productType;
    protected String productGroup;
    protected String tasksGroupName;
    protected Class<? extends BaseRunTask> runTaskClass;
    private String productName;
    private String prepareSandboxTaskName;
    private String patchPluginXmlTaskName;
    private String pluginXmlDirName;

    @Override
    public void apply(final Project project) {
        project.getPlugins().apply(JavaPlugin.class);
        final ProductPluginExtension extension = project.getExtensions()
                .create(extensionName, ProductPluginExtension.class);

        extension.setType(productType);
        extension.setPluginName(project.getName());
        File defaultSandboxDirectory = new File(project.getBuildDir(),productName.toLowerCase() + "-sandbox");
        extension.setSandboxDirectory(defaultSandboxDirectory);
        extension.setRepository(getRepositoryTemplate());
        extension.setExtensionProject(project);
        extension.setPlugin(this);
        extension.setIdePath(Utils.getDefaultIdePath(project, this, productType,
                ProductPluginExtension.DEFAULT_VERSION));

        configureTasks(project, extension);
    }

    protected abstract String getRepositoryTemplate();

    private void configureTasks(@NotNull Project project, @NotNull final ProductPluginExtension extension) {
        LOG.info("Configuring {} gradle plugin", productName);
        configurePatchPluginXmlTask(project, extension);
        configurePrepareSandboxTask(project, extension);
        configureRunTask(project, extension);
        configureProcessResources(project);
        project.afterEvaluate(it -> configureProjectAfterEvaluate(it, extension));
    }

    private void configureProjectAfterEvaluate(@NotNull Project project, @NotNull ProductPluginExtension extension) {
        for (Project subproject : project.getSubprojects()) {
            ProductPluginExtension subprojectExtension = (ProductPluginExtension) subproject.getExtensions()
                    .findByName(extensionName);
            if (subprojectExtension != null) {
                configureProjectAfterEvaluate(subproject, subprojectExtension);
            }
        }

        configureDependency(project, extension);
    }

    private void configureDependency(@NotNull Project project, @NotNull ProductPluginExtension extension) {
        LOG.info("Configuring {} dependency", productName);

        File idePath = extension.getIdePath();

        final String ideVersion = extension.getVersion();

        if (!idePath.exists()) {
            System.out.println("Start download " + productName + extension.getType() + "-" + ideVersion);

            File file = downloadProduct(extension, ideVersion);

            if (file == null) {
                System.out.println(productName + " not loaded");
                LOG.warn("{} not loaded from {}", productName, extension.getRepository());
                return;
            }

            System.out.println(productName + " Loaded");
            System.out.println("Start Unzip " + productName + "...");

            UnZipper.unZip(file, idePath);
            System.out.println("Unzipped " + productName + " to " + idePath);
        }

        ProductDependency dependency = DependencyManager.resolveLocal(project, extension, idePath, productName);
        extension.setDependency(dependency);
        DependencyManager.register(project, dependency, productName);

        File toolsJar = Jvm.current().getToolsJar();
        if (toolsJar != null) {
            project.getDependencies().add(JavaPlugin.RUNTIME_CONFIGURATION_NAME, project.files(toolsJar));
        }
    }

    @Nullable
    private File downloadProduct(@NotNull ProductPluginExtension extension, @NotNull String ideVersion) {
        URL url;
        File file;
        String repository = extension.getRepository();
        if (repository == null) {
            return null;
        }

        try {
            url = new URL(repository);
            Path tempDirectory = Files.createTempDirectory("product");
            file = tempDirectory.resolve(ideVersion + ".zip").toFile();

            try (BufferedInputStream bis = new BufferedInputStream(url.openStream())) {
                try (FileOutputStream fis = new FileOutputStream(file)) {
                    byte[] buffer = new byte[1024];
                    int count;
                    while ((count = bis.read(buffer, 0, 1024)) != -1) {
                        fis.write(buffer, 0, count);
                    }
                }
            }
        } catch (IOException e) {
            LOG.error("Failure download {} from {}", productName, repository);
            return null;
        }
        return file;
    }

    private void configurePrepareSandboxTask(
            @NotNull final Project project,
            @NotNull final ProductPluginExtension extension) {
        LOG.info("Configuring prepare {} sandbox task", productName);

        project.getTasks().create(prepareSandboxTaskName, PrepareSandboxTask.class,
                task -> {
                    task.setGroup(tasksGroupName);
                    task.setDescription("Prepare sandbox directory with installed plugin and its dependencies.");
                    task.setExtension(extension);
                    task.dependsOn(JavaPlugin.JAR_TASK_NAME);
                });
    }

    private void configurePatchPluginXmlTask(
            @NotNull final Project project,
            @NotNull final ProductPluginExtension extension) {
        LOG.info("Configuring patch plugin.xml task");

        project.getTasks().create(patchPluginXmlTaskName, PatchPluginXmlTask.class,
                task -> {
                    task.setGroup(tasksGroupName);
                    task.setDescription("Patch plugin xml files with corresponding since/until build numbers and version attributes");
                    task.setDestinationDirName(pluginXmlDirName);
                    task.setExtension(extension);
                });
    }

    private void configureRunTask(@NotNull final Project project, @NotNull final ProductPluginExtension extension) {
        LOG.info("Configuring run {} task", productName);

        project.getTasks().create("run" + productName,
                (Class<? extends BaseRunTask>) runTaskClass,
                (Action<BaseRunTask>) task -> {
                    task.setGroup(tasksGroupName);
                    task.setDescription("Runs " + productName + " with installed plugin.");
                    task.setExtension(extension);
                    task.setPlugin(this);

                    task.dependsOn(prepareSandboxTaskName);
                });
    }

    private void configureProcessResources(@NotNull Project project) {
        LOG.info("Configuring {} resources task", productName);
        ProcessResources task = (ProcessResources) Utils.findTask(project, JavaPlugin.PROCESS_RESOURCES_TASK_NAME);

        if (task != null) {
            PatchPluginXmlTask patchTask = (PatchPluginXmlTask) Utils.findTask(project, patchPluginXmlTaskName);
            task.from(patchTask, copySpec -> {
                copySpec.into(new File("META-INF"));
                copySpec.setDuplicatesStrategy(DuplicatesStrategy.INCLUDE);
            });
        }
    }

    String getProductName() {
        return productName;
    }

    protected void setProductName(final String productName) {
        this.productName = productName;
        this.prepareSandboxTaskName = "prepare" + productName + "Sandbox";
        this.patchPluginXmlTaskName = "patch" + productName + "PluginXml";
        this.pluginXmlDirName = "patched" + productName + "PluginXmlFiles";
    }

    String getProductGroup() {
        return productGroup;
    }

    String getPrepareSandboxTaskName() {
        return prepareSandboxTaskName;
    }

    String getProductType() {
        return productType;
    }
}

File: buildSrc/src/main/kotlin/org/stepik/gradle/plugins/common/BasePlugin.kt
package org.stepik.gradle.plugins.common

import org.gradle.api.IllegalDependencyNotation
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.file.DuplicatesStrategy.INCLUDE
import org.gradle.api.plugins.JavaPlugin
import org.gradle.api.plugins.JavaPlugin.JAR_TASK_NAME
import org.gradle.api.plugins.JavaPlugin.PROCESS_RESOURCES_TASK_NAME
import org.gradle.api.plugins.JavaPlugin.RUNTIME_ELEMENTS_CONFIGURATION_NAME
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.api.tasks.bundling.Zip
import org.gradle.api.tasks.testing.Test
import org.gradle.internal.jvm.Jvm
import org.gradle.language.jvm.tasks.ProcessResources
import org.slf4j.LoggerFactory
import org.stepik.gradle.plugins.common.RepositoryType.MAVEN
import org.stepik.gradle.plugins.common.Utils.getProductJvmArgs
import org.stepik.gradle.plugins.common.dependency.DependencyManager.register
import org.stepik.gradle.plugins.common.dependency.DependencyManager.resolveLocal
import org.stepik.gradle.plugins.common.dependency.DependencyManager.resolveLocalCashRepository
import org.stepik.gradle.plugins.common.dependency.DependencyManager.resolveRemoteMaven
import org.stepik.gradle.plugins.common.tasks.BaseRunTask
import org.stepik.gradle.plugins.common.tasks.InstrumentCodeTask
import org.stepik.gradle.plugins.common.tasks.PatchPluginXmlTask
import org.stepik.gradle.plugins.common.tasks.PrepareSandboxTask
import org.stepik.gradle.plugins.common.tasks.PublishTask
import java.io.File


abstract class BasePlugin(
        val extensionName: String,
        val productName: String,
        val productType: String,
        val productGroup: String,
        val tasksGroupName: String,
        val runTaskClass: Class<out BaseRunTask>,
        val extensionInstrumentCode: Boolean,
        val repositoryType: RepositoryType,
        val repositoryTemplate: String
) : Plugin<Project> {
    private val logger = LoggerFactory.getLogger(BasePlugin::class.java)

    val prepareSandboxTaskName = "prepare${productName}Sandbox"
    private val prepareTestSandboxTaskName = "prepare${productName}TestSandbox"
    private val patchPluginXmlTaskName = "patch${productName}PluginXml"
    val buildPluginTaskName = "build${productName}Plugin"

    override fun apply(project: Project) {
        project.plugins.apply(JavaPlugin::class.java)
        val extension = project.extensions
                .createProductPluginExtension(extensionName,
                        projectName = project.name,
                        productName = productName,
                        productType = productType,
                        productGroup = productGroup,
                        sandboxDirectory = File(project.buildDir, "${productName.toLowerCase()}-sandbox").toString(),
                        repository = repositoryTemplate,
                        project = project,
                        plugin = this,
                        instrumentCode = extensionInstrumentCode,
                        repositoryType = repositoryType,
                        publish = ProductPluginExtensionPublish()
                )

        configureTasks(project, extension)
    }

    private fun configureTasks(project: Project, extension: ProductPluginExtension) {
        logger.info("Configuring {} gradle plugin", productName)

        configurePatchPluginXmlTask(project, extension)
        configurePrepareSandboxTask(project, extension)
        configureRunTask(project, extension)
        configureBuildPluginTask(project)
        configurePublishPluginTask(project, extension)
        configureProcessResources(project)
        project.afterEvaluate {
            configureProjectAfterEvaluate(it, extension)
        }
    }

    private fun configureProjectAfterEvaluate(project: Project,
                                              extension: ProductPluginExtension) {
        project.subprojects.forEach {
            val subprojectExtension = it.extensions
                    .findByName(extensionName) as ProductPluginExtension? ?: return@forEach
            configureProjectAfterEvaluate(it, subprojectExtension)
        }

        configureDependency(project, extension)
        configureInstrumentation(project, extension)
        configureTestTasks(project, extension)
    }

    private fun configureDependency(project: Project, extension: ProductPluginExtension) {
        logger.info("Configuring {} dependency", productName)

        val dependency = if (extension.repositoryType == MAVEN) {
            resolveRemoteMaven(project, extension)
        } else {
            resolveLocalCashRepository(project, this, extension)
        } ?: resolveLocal(project, extension) ?: throw IllegalDependencyNotation()

        extension.dependency = dependency
        register(project, dependency, extension)

        val toolsJar = Jvm.current().toolsJar ?: return
        project.dependencies.add(RUNTIME_ELEMENTS_CONFIGURATION_NAME, project.files(toolsJar))
    }

    private fun configurePrepareSandboxTask(project: Project, ext: ProductPluginExtension) {
        logger.info("Configuring prepare {} sandbox task", productName)

        project.tasks.create(prepareSandboxTaskName, PrepareSandboxTask::class.java).apply {
            group = tasksGroupName
            description = "Prepare sandbox directory with installed plugin and its dependencies."
            extension = ext
            dependsOn(JAR_TASK_NAME)
        }
    }

    private fun configurePatchPluginXmlTask(project: Project, ext: ProductPluginExtension) {
        logger.info("Configuring patch plugin.xml task")

        project.tasks.create(patchPluginXmlTaskName, PatchPluginXmlTask::class.java).apply {
            group = tasksGroupName
            description = "Patch plugin xml files with corresponding since/until build numbers and version attributes"
            extension = ext
        }
    }

    private fun configureRunTask(project: Project, ext: ProductPluginExtension) {
        logger.info("Configuring run {} task", productName)

        project.tasks.create("run$productName", runTaskClass).apply {
            group = tasksGroupName
            description = "Runs $productName with installed plugin."
            extension = ext
            plugin = this@BasePlugin
            dependsOn(prepareSandboxTaskName)
        }
    }

    private fun configurePublishPluginTask(project: Project, ext: ProductPluginExtension) {
        logger.info("Configuring publishing {} plugin task", productName)
        project.tasks.create("publish$productName", PublishTask::class.java).apply {
            group = tasksGroupName
            description = "Publish plugin distribution on plugins.jetbrains.com."
            extension = ext
            plugin = this@BasePlugin
            dependsOn(buildPluginTaskName)
        }
    }

    private fun configureProcessResources(project: Project) {
        logger.info("Configuring {} resources task", productName)

        val processResourcesTask = project.tasks.findByName(PROCESS_RESOURCES_TASK_NAME) as ProcessResources?
        if (processResourcesTask != null) {
            val patchPluginXmlTask = project.tasks.findByName(patchPluginXmlTaskName) ?: return
            processResourcesTask.from(patchPluginXmlTask) {
                it.into("META-INF")
                it.duplicatesStrategy = INCLUDE
            }
        }
    }

    private fun configureInstrumentation(project: Project, extension: ProductPluginExtension) {
        logger.info("Configuring IntelliJ compile tasks")
        project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets.forEach {
            val name = it.getTaskName("${productName}Instrument", "code")
            val instrumentTask = project.tasks.create(name, InstrumentCodeTask::class.java).apply {
                this.sourceSet = it
                dependsOn(it.classesTaskName)
                onlyIf { extension.instrumentCode && extension.productType != "RS" }

                conventionMapping("ideaDependency", { extension.dependency })
                conventionMapping("outputDir", { File(it.output.classesDirs.first().parent, it.name) })
            }

            // Ensure that our task is invoked when the source set is built
            it.compiledBy(instrumentTask)
        }
    }

    private fun configureTestTasks(project: Project, extension: ProductPluginExtension) {
        logger.info("Configuring IntelliJ tests tasks")
        project.tasks.withType(Test::class.java).forEach {
            val configDirectory = project.file("${extension.sandboxDirectory}/config-test")
            val systemDirectory = project.file("${extension.sandboxDirectory}/system-test")
            val pluginsDirectory = project.file("${extension.sandboxDirectory}/plugins-test")

            it.enableAssertions = true
            it.systemProperties(extension.systemProperties)
            it.systemProperties(Utils.getProductSystemProperties(configDirectory, systemDirectory, pluginsDirectory))
            val idePath = extension.ideDirectory ?: return@forEach
            it.jvmArgs = getProductJvmArgs(it, it.jvmArgs, idePath)
            val dependency = extension.dependency
            if (dependency != null) {
                it.classpath += project.files("${dependency.classes}/lib/resources.jar",
                        "${dependency.classes}/lib/idea.jar")
            }
            it.outputs.dir(systemDirectory)
            it.outputs.dir(configDirectory)
            it.dependsOn(project.getTasksByName(prepareTestSandboxTaskName, false))
        }
    }

    private fun configureBuildPluginTask(project: Project) {
        logger.info("Configuring building plugin task")
        val prepareSandboxTask = project.tasks.findByName(prepareSandboxTaskName) as PrepareSandboxTask
        project.tasks.create(buildPluginTaskName, Zip::class.java).apply {
            description = "Bundles the project as a distribution."
            group = tasksGroupName
            from("${prepareSandboxTask.destinationDir}/${prepareSandboxTask.pluginName}")
            into { prepareSandboxTask.pluginName }
            dependsOn(prepareSandboxTask)
            baseName = prepareSandboxTask.pluginName
        }
    }

}

